### PR TITLE
Java buffer: Allow returned output length in a larger allocated fixed destination buffer.

### PR DESCRIPTION
**Pull request by cefaci**
Be sure to allocate enough bytes before (e.g. using the experimental ctc_beam_search_decoder in TFLite).

The check against the exact source buffer length with the exact destination buffer length is not useful in Java, even if you use it in the original way because then you know the exact expected returned length (you can still check in your code).

### CODE REVIEW

**Comment by cefaci** [comment]
@googlebot I signed it!

**Comment by teijeong** [comment]
Hi Jared, can you take a look?

File: tensorflow/lite/java/src/main/java/org/tensorflow/lite/Tensor.java [hunk 1]
```diff
@@ -420,7 +420,8 @@ private void throwIfShapeIsIncompatible(Object o) {
       // Note that we allow the client to provide a ByteBuffer even for non-byte Tensors.
       // In such cases, we only care that the raw byte capacity matches the tensor byte capacity.
       int oBytes = isByteBuffer(o) ? oBuffer.capacity() : oBuffer.capacity() * dtype.byteSize();
-      if (bytes != oBytes) {
+      // Allow dynamic output sizes, but be sure to allocate enough bytes in the destination
+      if (bytes > oBytes) {
```
Comment (jdduke) [inline tensorflow/lite/java/src/main/java/org/tensorflow/lite/Tensor.java:424]: What about the case where we're copying *from* an input buffer that is too large? I think we might need to distinguish between the cases where we're copying to/from an buffer?

**Review by jdduke — CHANGES_REQUESTED** [review]
Looks good but I think we need to consider the case where we're copying from an input buffer that is too large.

File: tensorflow/lite/java/src/main/java/org/tensorflow/lite/Tensor.java (hunk 1 above)
Comment (cefaci) [reply tensorflow/lite/java/src/main/java/org/tensorflow/lite/Tensor.java:424]: Is this not the original idea of this check and throwing this exception (which will still work for:
"_copying from an input buffer that is too large_")? 

I needed exactly this specific change as I don't know before the expected dynamic returned buffer size from the decoded characters found by ctc_beam_search_decoder. And in Java a buffer fits always in an _equal or larger_ output buffer, so I pre-allocated the result buffer always larger than the expected results can be.

Comment (jdduke) [reply tensorflow/lite/java/src/main/java/org/tensorflow/lite/Tensor.java:424]: There are 4 cases we need to consider:

1) The input (user) buffer is too large for the input *tensor*. - This should be an error.
2) The input buffer is too small for the input tensor - This should probably also be an error.
3) The output (user) buffer is too large for the output tensor - This is probably fine and is what you want to handle.
4) The output buffer is too small for the output tensor - This should be an error.

This method gets called for all 4 cases, but isn't reporting an error for (1).

Comment (cefaci) [reply tensorflow/lite/java/src/main/java/org/tensorflow/lite/Tensor.java:424]: Yes, true, you're right, checked it, it is used for both and as my input images were always right for my model, it worked... What do you recommend, copy the method and allow it only for _copyTo()_?

**Comment by gbaned** [comment]
@cefaci Can you please check @jdduke's comments and keep us posted. Thanks!

**Comment by jdduke** [comment]
I'm still thinking about how best to structure the code to handle this.

**Comment by cefaci** [comment]
How about passing an attribute (e.g. boolean) to Tensor.java over the NativeInterpreterWrapper.java to allow a dynamic output (> instead of !=)?

**Comment by cefaci** [comment]
Thanks! Looks good!

**Comment by jdduke** [comment]
Let us know if you run into any issues, and thanks for your patience!

**Comment by cefaci** [comment]
@jdduke I will definitely, no the other way round, thank you all for your commitment and time! :)

**Pull request closed without merging**